src/components/productrecommendation.fetchallpages.jsx

```diff
@@ -0,0 +1,163 @@
+          },
+        })
+        .then(res => res.json())
+        .then((res) => {
+          const arr = res.links;
+          for (let i = 0; i < arr.length; i += 1) {
+            if (arr[i].rel === 'element') {
+              link.push(arr[i]);
+            } else if (arr[i].rel === 'next') {
+              this.renderUrl(arr[i].uri);
+            }
+          }
+          this.setState({
+            links: link,
+          });
+          return res;
+        });
+    });
+  }
+
+  static renderProducts(product, length, MaxItemsInOneCarouselView) {
+    const totalCount = product.length;
+    const maxViews = totalCount / MaxItemsInOneCarouselView;
+    const data = [];
+    for (let CurrentView = 0, CurrentItem = 0; CurrentView < maxViews; CurrentView += 1, CurrentItem += MaxItemsInOneCarouselView) {
+      if (CurrentView === 0) {
+        data.push(
+          <div className="carousel-item active" key={(CurrentItem)}>
+            <div className="row">
+              {Productrecommendationfetchallpages.renderCarouselView(CurrentItem, totalCount, product, MaxItemsInOneCarouselView)}
+            </div>
+          </div>,
+        );
+      } else {
+        data.push(
+          <div className="carousel-item " key={(CurrentItem + 1)}>
+            <div className="row">
+              {Productrecommendationfetchallpages.renderCarouselView(CurrentItem, totalCount, product, MaxItemsInOneCarouselView)}
+            </div>
+          </div>,
+        );
+      }
+    }
+    return (
+      <div>
+        {data}
+      </div>
+    );
+  }
+
+  static renderCarouselView(currentItem, totalCount, product, MaxItemsInOneCarouselView) {
+    /* Renders Each View in Carousel */
+    const data = [];
+    let itemCount = currentItem;
+    if (itemCount < totalCount && product[itemCount].rel === 'element') {
+      data.push(
+        <div className="col-6" key={`_${Math.random().toString(36).substr(2, 9)}`}>
+          <ProductListItemMain productUrl={product[itemCount].uri} />
+        </div>,
+      );
+    }
+    for (let i = 1; i < MaxItemsInOneCarouselView; i++) {
+      /* Copy this section to add more items in the same carousel view */
+      itemCount += 1;
+      if ((itemCount) < totalCount && product[(itemCount)].rel === 'element') {
+        data.push(
+          <div className="col-6" key={`_${Math.random().toString(36).substr(2, 9)}`}>
+            <ProductListItemMain productUrl={product[(itemCount)].uri} />
+          </div>,
+        );
+      }
+    }
+
+    return data;
+  }
+
+  render() {
+    const data = [];
+    const { length, count } = this.props;
+    const { links } = this.state;
+    if (links.length > 0) {
+      data.push(
+        <div className="carousel-inner" key={`_${Math.random().toString(36).substr(2, 9)}`}>
+          {Productrecommendationfetchallpages.renderProducts(links, length, count)}
+        </div>,
+      );
+    }
+    return (
+      <div className="container">
+        {data}
```

Comment:
why bother returning an empty container if you don't have the data?

Reply:
this might be the cause of your isCancelled, your component won't actually unmount if there is no data, because you're using an empty container... if you return null when there is nothing I think this would help.

Reply:
also you can return an element conditionally like this:
{bool && <div etc./>}

Reply:
-- pending for discussion--